security/advancedtls/crl.go

```diff
@@ -0,0 +1,498 @@
+
+		}
+	}
+
+	// Finally, OpenSSL drops the initial sequence tag
+	// So, we marshal all the RDNs separately instead of as a group
+	for _, canonRdn := range r {
+		b, err := asn1.Marshal(canonRdn)
+		if err != nil {
+			grpclog.Print(err)
+		}
+		canonBytes = append(canonBytes, b...)
+	}
+
+	issuerHash := sha1.Sum(canonBytes)
+	// Openssl takes the first 4 bytes and encodes them as a little endian
+	// uint32 and then uses the hex to make the file name
+	// In C++, this would be:
+	// (((unsigned long)md[0]) | ((unsigned long)md[1] << 8L) |
+	// ((unsigned long)md[2] << 16L) | ((unsigned long)md[3] << 24L)
+	// ) & 0xffffffffL;
+	fileHash := binary.LittleEndian.Uint32(issuerHash[0:4])
+	return fmt.Sprintf("%08x", fileHash)
+}
+
+// CheckRevocation checks the connection for revoked certificates based on RFC5280.
+// This implementation has the following major limitations:
+// 	Indirect CRL files are not supported
+// 	File based lookup only using the X509_LOOKUP_hash_dir format
+// 	OnlySomeReasons is unsupported
+// 	Delta CRL files are unsupported
+// 	Certificate cRLDistributionPoint must be URLs, but are then ignored and converted into a file path
+// 	CRL checks are done after path building, which goes against rfc4158
+func CheckRevocation(conn tls.ConnectionState, cfg RevocationConfig) error {
+	return CheckChainRevocation(conn.VerifiedChains, cfg)
+}
+
+// CheckChainRevocation checks the verified certificate chain
+// for revoked certificates based on RFC5280.
+func CheckChainRevocation(verifiedChains [][]*x509.Certificate, cfg RevocationConfig) error {
+	// Iterate the verified chains looking for one that is RevocationUnrevoked.
+	// A single RevocationUnrevoked chain is enough to allow the connection, and a single RevocationRevoked
+	// chain does not mean the connection should fail
+	count := make(map[RevocationStatus]int)
+	for _, chain := range verifiedChains {
+		switch checkChain(chain, cfg) {
+		case RevocationUnrevoked:
+			// If any chain is RevocationUnrevoked then return no error
+			return nil
+		case RevocationRevoked:
+			// If this chain is revoked, keep looking for another chain
+			count[RevocationRevoked]++
+			continue
+		case RevocationUndetermined:
+			if cfg.AllowUndetermined {
+				return nil
+			}
+			count[RevocationUndetermined]++
+			continue
+		}
+	}
+	return fmt.Errorf("no unrevoked chains found: %v", count)
+}
+
+// checkChain will determine and check all certificates in chain against the CRL
+// defined in the certificate with the following rules:
+// 1. If any certificate is RevocationRevoked, return RevocationRevoked
+// 2. If any certificate is RevocationUndetermined, return RevocationUndetermined
+// 3. If all certificates are RevocationUnrevoked, return RevocationUnrevoked
+func checkChain(chain []*x509.Certificate, cfg RevocationConfig) RevocationStatus {
+	chainStatus := RevocationUnrevoked
+	for _, c := range chain {
+		switch checkCrt(c, chain, cfg) {
```

Comment:
Is `Crt` short for "certificate"?  If so, maybe use `Cert`, which is a more common abbreviation.